src/Dictionaries/CacheDictionary.cpp

```diff
@@ -822,7 +822,24 @@ void CacheDictionary::waitForCurrentUpdateFinish(UpdateUnitPtr & update_unit_ptr
 
 
     if (update_unit_ptr->current_exception)
-        std::rethrow_exception(update_unit_ptr->current_exception);
+    {
+        // There might have been a single update unit for multiple callers in
+        // independent threads, and current_exception will be the same for them.
+        // Don't just rethrow it, because sharing the same exception object
+        // between multiple threads can lead to weird effects if they decide to
+        // modify it, for example, by adding some error context.
```

Comment:
But this is not true. I've deleted the code of glueing multiple update units into a bigger one here #13624. Because of several problems, in particular because of this.

Reply:
If you run 01412_cache_dictionary_race, and print the address of exception here: https://github.com/ClickHouse/ClickHouse/pull/14689/files#diff-8e99551591f6234b1c0996129183d6f7R617
you will see that it is the same exception object. Better add flush + sleep after the print, so that it is more clear. And after that you will see the TSan failure that tells the same thing.

After this change it stopped reproducing locally for me.

Maybe I'm misinterpreting the reason, e.g. there's some deeper mechanism that leads to this?